src/opm/input/eclipse/EclipseState/Grid/Carfin.cpp

```diff
@@ -0,0 +1,292 @@
+        default_count += update_default(deckRecord.getItem<ParserKeywords::CARFIN::NY>(), ny);
+        default_count += update_default(deckRecord.getItem<ParserKeywords::CARFIN::NZ>(), nz);
+
+        if (default_count != 10) {
+            this->init(name, i1, i2, j1, j2, k1, k2, nx, ny, nz);
+        }
+    }
+
+    void Carfin::reset()
+    {
+        this->init("LGR", 0, this->m_globalGridDims_.getNX() - 1,
+                   0, this->m_globalGridDims_.getNY() - 1,
+                   0, this->m_globalGridDims_.getNZ() - 1,
+                   this->m_globalGridDims_.getNX(), this->m_globalGridDims_.getNY(), this->m_globalGridDims_.getNZ());
+    }
+
+    void Carfin::init(const std::string name, const int i1, const int i2,
+                   const int j1, const int j2,
+                   const int k1, const int k2,
+                   const int nx, const int ny,
+                   const int nz)
+    {
+        assert_dims(name, i1 , i2, nx, this->m_globalGridDims_.getNX());
+        assert_dims(name, j1 , j2, ny, this->m_globalGridDims_.getNY());
+        assert_dims(name, k1 , k2, nz, this->m_globalGridDims_.getNZ());
+
+        this->m_dims[0] = nx;  
+        this->m_dims[1] = ny;
+        this->m_dims[2] = nz;
+
+        this->m_offset[0] = static_cast<std::size_t>(i1);
+        this->m_offset[1] = static_cast<std::size_t>(j1);
+        this->m_offset[2] = static_cast<std::size_t>(k1);
+
+        this->m_end_offset[0] = static_cast<std::size_t>(i2);
+        this->m_end_offset[1] = static_cast<std::size_t>(j2);
+        this->m_end_offset[2] = static_cast<std::size_t>(k2);
+
+        this->initIndexList();
+    }
+
+    std::size_t Carfin::size() const
+    {
+        return m_dims[0] * m_dims[1] * m_dims[2];
+    }
+
+    bool Carfin::isGlobal() const
+    {
+        return this->size() == this->m_globalGridDims_.getCartesianSize();
+    }
+
+    std::size_t Carfin::getDim(std::size_t idim) const
+    {
+        if (idim >= 3) {
+            throw std::invalid_argument("The input dimension value is invalid");
+        }
+
+        return m_dims[idim];
+    }
+
+    const std::vector<Carfin::cell_index>& Carfin::index_list() const {
+        return this->m_active_index_list;
+    }
+
+    const std::vector<Carfin::cell_index>& Carfin::global_index_list() const {
+        return this->m_global_index_list;
+    }
+
+    void Carfin::initIndexList() 
+    {
+        this->m_active_index_list.clear();
+        this->m_global_index_list.clear();
+
+        const auto lgrdims = GridDims(this->m_dims[0], this->m_dims[1], this->m_dims[2]);
+        const auto ncells = lgrdims.getCartesianSize();
+
+        for (auto data_index = 0*ncells; data_index != ncells; ++data_index) {
+            const auto lgrIJK = lgrdims.getIJK(data_index);
+            const auto global_index = this->m_globalGridDims_
+                .getGlobalIndex(lgrIJK[0]/this->m_dims[0]*(this->m_end_offset[0]-this->m_offset[0]+1) + this->m_offset[0], 
```

Comment:
Let's skip the multiplication with `this->m_end_offset[i]-this->m_offset[i]+1`. Should be 1 all the time